gempy/share/man/man1/dataselect.1
.TH dataselect 1 "Oct 2019" "version 0.1" "dataselect man page"
.SH NAME
dataselect
.SH SYNOPSIS
.B dataselect
[\fB\-h\fR]
[\fB\-t\fR TAGS [TAGS ...], \fB\-\-tags\fR TAGS [TAGS ...]]
[\fB\-\-xtags\fR XTAGS [XTAGS ...]]
[\fB\-\-expr\fR EXPRESSION]
[\fB-\-strict\fR]
[\fB\-o\fR OUTPUT, \fB\-\-output\fR OUTPUT]
[\fB\-v\ , \fB\-\-verbose\fR]
[\fB\-\-debug\fR]
inputs [inputs ...]

.SH DESCRIPTION
.B dataselect
helps users find files matching certain criteria defined by astrodata tags and
expressions involving descriptors. Descriptors are specified to fine tune
dataselect matches.

As the \fB--tags\fR option indicates, \fBdataselect\fR can find and report data
that match specific tag criteria. For example, a user might want to find all
GMOS image flats under a certain directory. \fBdataselect\fR will locate and
report all datasets that would match the tag set (GMOS, IMAGE, FLAT).

A user may request that a file be written containing all datasets matching
the data selection. An output file is specified through the \fB\-o\fR,
\fB\-\-output\fR option. Output files are formatted so they may be passed
directly to the reduce command line via that applications 'at-file' (@file)
facility. (See the \fIRecipe System Users Manual\fR for further detail.)

.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
show the \fBdataselect\fR help message and exit
.TP
\fB\-t\fR TAG [,TAG ...], \fB\-\-tags\fR TAG [,TAG ...]
Comma-separated list of required tags. Find datasets that match only these tag
criteria.

.RS
Eg.,
.RS
\fB\-\-tags\fR SOUTH,GMOS,IMAGE

.RE
will report datasets that are all tagged \fBSOUTH\fR and \fBGMOS\fR and \fBIMAGE\fR.
.RE
.TP
\fB\-\-xtags\fR XTAG [,XTAG ...]
Comma-separated list of tags to exclude from reporting. Same form as
.B -t, --tags
option (above).
.TP
\fB\-\-expr\fR EXPRESSION
Show files matching <EXPRESSION>.
.TP
.TP
\fB\-\-strict
Toggle on strict expression matching for
.B exposure_time
(not just close) and for
.B filter_name
(match component number).
.TP
\fB\-o\fR OUTPUT, \fB\-\-output\fR OUTPUT
Write reported files to this file. Effective only with \fB\-\-tags\fR option.
.TP
\fB\-v\fR, \fB-\-verbose\fR
Toggle verbose mode when using [-o, --output]. Default is False.
.TP
\fB\-\-debug\fR
Toggle debug mode. Default is False.
.SH EXAMPLES
Select all the FITS files inside the raw directory with a tag that matches DARK.

.RS
$ dataselect raw/*.fits --tags DARK

.RE
To select darks of a specific exposure time:

.RS
$ dataselect raw/*.fits --tags DARK --expr='exposure_time==20'
.RE

To send that list to a file that can be used later:

.RS
$ dataselect raw/*.fits --tags DARK --expr='exposure_time==20' -o dark20s.lis
.RE

.SH ENVIRONMENT
Recommend Anaconda 4.2.0, Astropy 1.2.1.
.SH BUGS
None known.

.SH SEE ALSO
.I reduce(1), typewalk(1)


Changes.md
# 2.1.1

## Bug Fixes

**geminidr.core**
* Fix a crash when a section was used when stacking.

**gempy scripts**
* Add missing third party adpkg and drpkg support to utility scripts dataselect, showpars, typewalk, and showrecipes.

**gempy.library**
* fix to Jacobian calculation for non-affine transforms 

**recipe_system.adcc**
* Make adcc more robust to missing connection to fitsstore.


## Compatibility

**gempy.gemini**
* Add compatibility with sigma_clip for astropy v3.1+
* Add IRAF compatibility keywords on GMOS mosaiced data.
* Add compatibility with astroquery 0.4.

**geminidr.core**
* Add compatibility with sigma_clip fro astropy v3.1+ 
  
**geminidr.gmos**
* Add IRAF compatibility recipe.


## Documentation

* Various fixes to documentation and instruction manual following feedback from users.

gempy/adlibrary/dataselect.py
import re
from copy import deepcopy
from datetime import datetime  # leave it in.  It is used by the eval.

from importlib import import_module

import astrodata
import gemini_instruments


def isclose(a, b, rel_tol=1e-02, abs_tol=0.0):
    """

    Parameters
    ----------
    a : int/float
        first value
    b : int/float
        second value
    rel_tol : float, (optional)
        relative tolerance
    abs_tol : float, (optional)
        the absolute tolerance

    Returns
    -------

    """
    return abs(a - b) <= max(rel_tol * max(abs(a), abs(b)), abs_tol)

def expr_parser(expression, strict=False):
    """
    Takes a selection expression and return a codified version of that
    expression that returns True or False when eval().

    Parameters
    ----------
    expression : str
        the string that is parsed by the function
    strict : Bool
        Defines if expr_parser is "strict" or not

    Returns
    -------
    String that returns True or False when eval()

    """

    adexpr = re.sub(r'([_A-z]\w*)([!=<>]+\S+)', r'ad.\1()\2', expression)
    codified_expression = ''
    components = re.split(r'(\s+and\s+|\s+or\s+|\s+not\s+)', adexpr)

    for item in components:
        match = re.match(r'(ad.)([_A-z]\w*)([\(\)]+)([!=<>]+)(\S+)', item)
        if match:
            descriptor = match.groups()[1]
            operator = match.groups()[3]
            pattern = r'(ad.' + re.escape(
                descriptor) + r')([\(\)]+)([!=<>]+)(\S+)'
            if descriptor in ['ut_time', 'local_time']:
                codified_item = \
                    re.sub(pattern,
                           r'\1\2\3datetime.strptime(\4, "%H:%M:%S").time()',
                           item)
            elif descriptor == 'ut_date':
                codified_item = \
                    re.sub(pattern,
                           r'\1\2\3datetime.strptime(\4, "%Y-%m-%d").date()',
                           item)
            elif descriptor == 'ut_datetime':
                codified_item = \
                    re.sub(pattern,
                           r'\1\2\3datetime.strptime(\4, "%Y-%m-%d %H:%M:%S")',
                           item)
            elif descriptor == 'exposure_time' and operator == '==' and not strict:
                codified_item = \
                    re.sub(pattern, r'isclose(\1(),\4)', item)
            elif descriptor == 'central_wavelength' and operator == '==' and \
                not strict:
                codified_item = \
                    re.sub(pattern, r'isclose(\1(),\4, rel_tol=1e-5)', item)
            elif descriptor in ['filter_name', 'detector_name', 'disperser'] \
                and not strict:
                codified_item = re.sub(pattern, r'\1(pretty=True)\3\4', item)
            else:
                codified_item = item

            codified_expression += codified_item
        else:
            codified_expression += item

    return codified_expression


def evalexpression(ad, expression):
    """
    Tests to make sure that the expression passed from the user is a Bool
    Parameters
    ----------
    ad : astrodata.object
        Unused, remove if necessary
    expression : str
        expression from user

    Returns
    -------
    expression - bool
        evaluated expression, True or False
    """
    if type(eval(expression)) is not type(True):
        raise IOError('Expression does not return a boolean value.')
    return eval(expression)


def select_data(inputs, tags=[], xtags=[], expression='True', adpkg=None):
    """
    Given a list of fits files, function will return a list of astrodata objects
    that satisfy all the parameters

    Parameters
    ----------
    inputs - list of strings
        list of paths with full directory to a fits file
    tags - list of strings
        list with all the tags you want to make sure are in the inputs fits files
    xtags - list of strings
        list with all the tags you DO NOT want in the astrodata objects
    expression - usually dataselect.expr_parser object
        filters for any specific constraining like data,
        datetime, filter name, exposure time

    Returns
    list of astrodata objects
        returns all objects that satisfy all the above requirements
    -------
    """
    if adpkg is not None:
        import_module(adpkg)

    selected_data = []
    for input in inputs:
        ad = astrodata.open(input)
        adtags = ad.tags
        if set(tags).issubset(adtags) and \
               not len(set(xtags).intersection(adtags)) and \
               evalexpression(ad, expression):
           selected_data.append(input)

    return selected_data

def writeheader(fh, tags, xtags, expression):
    """
    Given a list of fits files, function will return a list of astrodata objects
    that satisfy all the parameters

    Parameters
    ----------
    fh -
        opened file
    tags - list of strings
        list with all the tags you want to make sure are in the inputs fits files
    xtags - list of strings
        list with all the tags you DO NOT want in the astrodata objects
    expression - usually dataselect.expr_parser object
        filters for any specific constraining like data,
        datetime, filter name, exposure time

    Returns
    N/a
    -------
    """
    if tags is None:
        tags = 'None'
    if xtags is None:
        xtags = 'None'
    if expression is None:
        expression = 'None'
    fh.write('# Includes tags: '+str(tags)+'\n')
    fh.write('# Excludes tags: '+str(xtags)+'\n')
    fh.write('# Descriptor expression: '+expression+'\n')
    return
